kotlin/workflow-core/src/main/java/com/squareup/workflow/WorkflowAction.kt

```diff
@@ -18,24 +18,51 @@
 package com.squareup.workflow
 
 import com.squareup.workflow.WorkflowAction.Companion.toString
-import com.squareup.workflow.WorkflowAction.Mutator
+import com.squareup.workflow.WorkflowAction.Updater
 
 /**
- * A function that can change the current state of a [Workflow] by returning a new one, and
- * also optionally emit an output.
+ * An atomic operation that updates the state of a [Workflow], and also optionally emits an output.
  */
-interface WorkflowAction<StateT, out OutputT : Any> {
+interface WorkflowAction<StateT, OutputT : Any> {
+  @Deprecated("Use Updater")
   class Mutator<S>(var state: S)
 
+
   /**
-   * Called with the current state of a workflow, allowing the state to be updated.
+   * The context for calls to [WorkflowAction.apply]. Allows the action to set the
+   * [nextState], and to emit the [nextOutput].
    *
-   * @return an output value for the workflow to emit, or null to emit nothing.
+   * @param nextState the state that the workflow should move to. Default is the current state.
    */
-  fun Mutator<StateT>.apply(): OutputT?
+  class Updater<S, O : Any>(var nextState: S) {
+    private var _nextOutput: O? = null
 
-  companion object {
+    /**
+     * The next output value the workflow should emit, or `null` to emit nothing.
+     * Defaults to `null`.
+     */
+    var nextOutput: O?
+      get() = _nextOutput
+      set(value) {
+        _nextOutput = requireNotNull(value)
+      }
+  }
 
+  @Suppress("DEPRECATION")
+  @Deprecated("Implement Updater.apply")
+  fun Mutator<StateT>.apply(): OutputT? {
+    throw UnsupportedOperationException()
```

Comment:
If this is just going to throw anyway, should the deprecation level be `Error`? If it's always going to fail anyway, might as well fail at compile time. But in that case I'm not sure what the point of leaving this method around is, since it will still require one giant pass to update.

Reply:
It won't require a single pass. If you implement your own Updater.apply() then this method will never be called. Nothing calls it except the default Updater.apply(). Existing actions will work unchanged. (I'll double check that, leave one of the sample unconverted and just suppress its warnings.)

Reply:
Confirmed.